tests/subsys/net/lib/downloader/src/main.c

```diff
@@ -0,0 +1,2162 @@
+
+	return 0;
+}
+
+static int dlc_callback_abort(const struct downloader_evt *event)
+{
+	size_t written;
+
+	TEST_ASSERT(event != NULL);
+
+	printk("event: %s\n", dlc_event_id_str(event->id));
+	if (event->id == DOWNLOADER_EVT_ERROR) {
+		/* avoid spamming error events during development */
+		k_sleep(K_MSEC(100));
+	}
+	k_pipe_put(&event_pipe, (void *)event, sizeof(*event), &written, sizeof(*event), K_FOREVER);
+
+	return 1; /* stop download*/
+}
+
+static struct downloader_evt dlc_wait_for_event(enum downloader_evt_id event,
+						     k_timeout_t timeout)
+{
+	size_t read;
+	struct downloader_evt evt;
+	int err;
+
+	while (true) {
+		err = k_pipe_get(&event_pipe, &evt, sizeof(evt), &read, sizeof(evt),
+				 timeout);
+		TEST_ASSERT_EQUAL(0, err);
+		if (evt.id == event) {
+			break;
+		}
+	}
+	TEST_ASSERT_EQUAL(evt.id, event);
+	return evt;
+}
+
+void test_downloader_init_einval(void)
+{
+	int err;
+	char buf[1];
+	struct downloader dlc = {};
+	struct downloader_cfg config = {
+		.buf = buf,
+		.buf_size = ARRAY_SIZE(buf),
+		.callback = dlc_callback,
+	};
+	struct downloader_cfg config_no_buf = {
+		.buf = NULL,
+		.buf_size = ARRAY_SIZE(buf),
+		.callback = dlc_callback,
+	};
+	struct downloader_cfg config_no_buf_len = {
+		.buf = buf,
+		.buf_size = 0,
+		.callback = dlc_callback,
+	};
+	struct downloader_cfg config_no_cb = {
+		.buf = buf,
+		.buf_size = ARRAY_SIZE(buf),
+		.callback = NULL,
+	};
+
+	err = downloader_init(NULL, &config);
+	TEST_ASSERT_EQUAL(-EINVAL, err);
+
+	err = downloader_init(&dlc, NULL);
+	TEST_ASSERT_EQUAL(-EINVAL, err);
+
+	err = downloader_init(&dlc, &config_no_buf);
+	TEST_ASSERT_EQUAL(-EINVAL, err);
+
+	err = downloader_init(&dlc, &config_no_buf_len);
+	TEST_ASSERT_EQUAL(-EINVAL, err);
+
+	err = downloader_init(&dlc, &config_no_cb);
+	TEST_ASSERT_EQUAL(-EINVAL, err);
+}
+
+void test_downloader_deinit_einval(void)
+{
+	int err;
+	struct downloader dl = {};
+
+	err = downloader_deinit(NULL);
+	TEST_ASSERT_EQUAL(-EINVAL, err);
+
+	dl.state = 0xffffffff;
+	/* bad state */
+	err = downloader_deinit(&dl);
+	TEST_ASSERT_EQUAL(-EINVAL, err);
+}
+
+void test_downloader_get_eperm(void)
+{
+	int err;
+
+	err = downloader_get(&dlc, &dlc_host_cfg, HTTP_URI, 0);
+	TEST_ASSERT_EQUAL(-EPERM, err);
+}
+
+void test_downloader_get_http(void)
+{
+	int err;
+	struct downloader_evt evt;
+
+	err = downloader_init(&dlc, &dlc_config);
+	TEST_ASSERT_EQUAL(0, err);
+
+	zsock_getaddrinfo_fake.custom_fake = zsock_getaddrinfo_server_ipv6_fail_ipv4_ok;
+	zsock_freeaddrinfo_fake.custom_fake = zsock_freeaddrinfo_server_ipv4;
+	z_impl_zsock_socket_fake.custom_fake = z_impl_zsock_socket_http_ipv4_ok;
+	z_impl_zsock_connect_fake.custom_fake = z_impl_zsock_connect_ipv4_ok;
+	z_impl_zsock_setsockopt_fake.custom_fake = z_impl_zsock_setsockopt_http_ok;
+	z_impl_zsock_sendto_fake.custom_fake = z_impl_zsock_sendto_ok;
+	z_impl_zsock_recvfrom_fake.custom_fake = z_impl_zsock_recvfrom_http_header_then_data;
```

Comment:
I don't know if there is something why CMock doesn't work in these tests but in general the tests are more readable with it and no need to write own mocks for simple things like here.

Reply:
I wasn't able to create proper mocks for the socket functions in zephyr with cmock (`socket`, `connect`, `recv`, `send`, ...) because of the way they are defined, though it might be that I just didn't find the right way to do it.

Reply:
Ok, no need to worry about this now.